src/computability/reduce.lean

```diff
@@ -0,0 +1,358 @@
+
+theorem many_one_equiv.congr_left {α β γ} [primcodable α] [primcodable β] [primcodable γ]
+  {p : α → Prop} {q : β → Prop} {r : γ → Prop}
+  (h : many_one_equiv p q) : many_one_equiv p r ↔ many_one_equiv q r :=
+and_congr h.le_congr_left h.le_congr_right
+
+theorem many_one_equiv.congr_right {α β γ} [primcodable α] [primcodable β] [primcodable γ]
+  {p : α → Prop} {q : β → Prop} {r : γ → Prop}
+  (h : many_one_equiv q r) : many_one_equiv p q ↔ many_one_equiv p r :=
+and_congr h.le_congr_right h.le_congr_left
+
+theorem one_one_equiv.congr_left {α β γ} [primcodable α] [primcodable β] [primcodable γ]
+  {p : α → Prop} {q : β → Prop} {r : γ → Prop}
+  (h : one_one_equiv p q) : one_one_equiv p r ↔ one_one_equiv q r :=
+and_congr h.le_congr_left h.le_congr_right
+
+theorem one_one_equiv.congr_right {α β γ} [primcodable α] [primcodable β] [primcodable γ]
+  {p : α → Prop} {q : β → Prop} {r : γ → Prop}
+  (h : one_one_equiv q r) : one_one_equiv p q ↔ one_one_equiv p r :=
+and_congr h.le_congr_right h.le_congr_left
+
+-- local attribute [instance] many_one_equiv_setoid
+
+/-- A many-one degree is an equivalence class of sets up to many-one equivalence. -/
+def many_one_degree (α) [primcodable α] := quotient (@many_one_equiv_setoid α _)
+
+instance {α} [primcodable α] : inhabited (many_one_degree α) :=
+⟨@quotient.mk _(@many_one_equiv_setoid α _) ∅⟩
+
+local infix ` ⊕' `:1001 := sum.elim
+
+open nat.primrec
+
+theorem one_one_reducible.disjoin_left {α β} [primcodable α] [primcodable β]
+  {p : α → Prop} {q : β → Prop} : p ≤₁ p ⊕' q :=
+⟨sum.inl, computable.sum_inl, λ x y, sum.inl.inj_iff.1, λ a, iff.rfl⟩
+
+theorem one_one_reducible.disjoin_right {α β} [primcodable α] [primcodable β]
+  {p : α → Prop} {q : β → Prop} : q ≤₁ p ⊕' q :=
+⟨sum.inr, computable.sum_inr, λ x y, sum.inr.inj_iff.1, λ a, iff.rfl⟩
+
+theorem disjoin_many_one_reducible {α β γ} [primcodable α] [primcodable β] [primcodable γ]
+  {p : α → Prop} {q : β → Prop} {r : γ → Prop} : p ≤₀ r → q ≤₀ r → p ⊕' q ≤₀ r
+| ⟨f, c₁, h₁⟩ ⟨g, c₂, h₂⟩ := ⟨sum.elim f g,
+  computable.id.sum_cases (c₁.comp computable.snd).to₂ (c₂.comp computable.snd).to₂,
+  λ x, by cases x; [apply h₁, apply h₂]⟩
+
+theorem disjoin_le {α β γ} [primcodable α] [primcodable β] [primcodable γ]
+  {p : α → Prop} {q : β → Prop} {r : γ → Prop} : p ⊕' q ≤₀ r ↔ p ≤₀ r ∧ q ≤₀ r :=
+⟨λ h, ⟨one_one_reducible.disjoin_left.to_many_one.trans h,
+  one_one_reducible.disjoin_right.to_many_one.trans h⟩,
+ λ ⟨h₁, h₂⟩, disjoin_many_one_reducible h₁ h₂⟩
+
+/--
+For many-one degrees `d₁` and `d₂`, `d₁ ≤ d₂` if the sets in `d₁` are many-one reducible to the
+sets in `d₂`.
+-/
+def many_one_degree.le {α β} [primcodable α] [primcodable β]
+  (d₁ : many_one_degree α) (d₂ : many_one_degree β) : Prop :=
+quotient.lift_on₂' d₁ d₂ (λ a b, a ≤₀ b)
+  (λ a b c d h₁ h₂, propext $ (h₁.le_congr_left).trans (h₂.le_congr_right))
+
+instance many_one_degree.has_le {α} [primcodable α] : has_le (many_one_degree α) := ⟨many_one_degree.le⟩
```

Comment:
I'm not sure if it is a good idea to have two ways to compare degrees (i.e., `.le` and `≤`).  Unfortunately the `.le` is more general than `≤`.